File: LeetCode/Medium/0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.java
class RandomizedSet {
    Map<Integer, Integer> locations;
    List<Integer> values;
    int size;

    public RandomizedSet() {
        locations = new HashMap<>();
        values = new ArrayList<>();
        size = 0;
    }
    
    public boolean insert(int val) {
        if (locations.containsKey(val)) {
            return false;
        }

        locations.put(val, size);
        values.add(val);
        size += 1;

        return true;
    }
    
    public boolean remove(int val) {
        if (!locations.containsKey(val)) {
            return false;
        }
        int loc = locations.get(val);
        int lastVal = values.get(size-1);

        Collections.swap(values, loc, size-1);
        locations.put(lastVal, loc);

        locations.remove(val);
        values.remove(size-1);
        size -= 1;

        return true;
    }
    
    public int getRandom() {
        return values.get(new Random().nextInt(values.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: LeetCode/Medium/0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.go
type RandomizedSet struct {
    set map[int]int
    keys []int
}


func Constructor() RandomizedSet {
    return RandomizedSet{
        set: make(map[int]int),
        keys: make([]int, 0),
        }
}


func (this *RandomizedSet) Insert(val int) bool {
    if _, exists := this.set[val]; exists {
        return false
    }
    this.set[val] = len(this.keys)
    this.keys = append(this.keys, val)
    return true
}


func (this *RandomizedSet) Remove(val int) bool {
    if _, exists := this.set[val]; !exists {
        return false
    }

    removeIdx := this.set[val]
    LastIdx := len(this.keys) - 1
    LastElem := this.keys[LastIdx]

    this.set[LastElem] = removeIdx
    this.keys[LastIdx], this.keys[removeIdx] = this.keys[removeIdx], this.keys[LastIdx]
    this.keys = this.keys[:LastIdx]

    delete(this.set, val)
    
    return true
}


func (this *RandomizedSet) GetRandom() int {
    randomIdx := rand.Intn(len(this.keys))
    return this.keys[randomIdx]
}


/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */